File: sdks/typescript/src/transactions.ts
import { SynapClient } from './client';
import { KVStore } from './kv';
import { HashManager } from './hash';
import { ListManager } from './list';
import { SetManager } from './set';
import { SortedSetManager } from './sorted-set';
import { QueueManager } from './queue';
import { StreamManager } from './stream';
import { PubSubManager } from './pubsub';
import { ScriptManager } from './scripting';
import { HyperLogLogManager } from './hyperloglog';

export interface TransactionOptions {
  /** Optional client identifier to group commands within the same transaction */
  clientId?: string;
}

export interface TransactionWatchOptions extends TransactionOptions {
  /** Keys to watch for changes */
  keys: string[];
}

export interface TransactionResponse {
  /** Informational message returned by the server */
  message: string;
  /** Indicates whether the operation succeeded */
  success: boolean;
}

export interface TransactionExecSuccess<T = unknown> {
  success: true;
  results: T[];
}

export interface TransactionExecAborted {
  success: false;
  aborted: true;
  message?: string;
}

export type TransactionExecResult<T = unknown> =
  | TransactionExecSuccess<T>
  | TransactionExecAborted;

export interface TransactionScope {
  clientId: string;
  kv: KVStore;
  hash: HashManager;
  list: ListManager;
  set: SetManager;
  sortedSet: SortedSetManager;
  queue: QueueManager;
  stream: StreamManager;
  pubsub: PubSubManager;
  script: ScriptManager;
  hyperloglog: HyperLogLogManager;
}

/**
 * Transaction manager for Redis-compatible MULTI/EXEC workflow.
 *
 * ```typescript
 * const client = new Synap({ url: 'http://localhost:15500' });
 * const txClientId = crypto.randomUUID();
 *
 * await client.transaction.multi({ clientId: txClientId });
 * await client.kv.set('counter', 1, { clientId: txClientId });
 * const result = await client.transaction.exec({ clientId: txClientId });
 * ```
 */
export class TransactionManager {
  constructor(private readonly client: SynapClient) {}

  private buildPayload(
    options?: TransactionOptions,
    extra: Record<string, unknown> = {}
  ): Record<string, unknown> {
    const payload: Record<string, unknown> = { ...extra };

    if (options?.clientId) {
      payload.client_id = options.clientId;
    }

    return payload;
  }

  /**
   * Start a new transaction (MULTI)
   */
  async multi(options?: TransactionOptions): Promise<TransactionResponse> {
    const response = await this.client.sendCommand<TransactionResponse>(
      'transaction.multi',
      this.buildPayload(options)
    );

    return {
      success: response.success ?? true,
      message: response.message ?? 'Transaction started',
    };
  }

  /**
   * Discard the current transaction (DISCARD)
   */
  async discard(options?: TransactionOptions): Promise<TransactionResponse> {
    const response = await this.client.sendCommand<TransactionResponse>(
      'transaction.discard',
      this.buildPayload(options)
    );

    return {
      success: response.success ?? true,
      message: response.message ?? 'Transaction discarded',
    };
  }

  /**
   * Watch keys for optimistic locking (WATCH)
   */
  async watch(options: TransactionWatchOptions): Promise<TransactionResponse> {
    if (!options.keys || options.keys.length === 0) {
      throw new Error('Transaction watch requires at least one key');
    }

    const response = await this.client.sendCommand<TransactionResponse>(
      'transaction.watch',
      this.buildPayload(options, { keys: options.keys })
    );

    return {
      success: response.success ?? true,
      message: response.message ?? 'Keys watched',
    };
  }

  /**
   * Remove all watched keys (UNWATCH)
   */
  async unwatch(options?: TransactionOptions): Promise<TransactionResponse> {
    const response = await this.client.sendCommand<TransactionResponse>(
      'transaction.unwatch',
      this.buildPayload(options)
    );

    return {
      success: response.success ?? true,
      message: response.message ?? 'Keys unwatched',
    };
  }

  /**
   * Execute queued commands (EXEC)
   */
  async exec<T = unknown>(options?: TransactionOptions): Promise<TransactionExecResult<T>> {
    const response = await this.client.sendCommand<Record<string, unknown>>(
      'transaction.exec',
      this.buildPayload(options)
    );

    if (Array.isArray(response?.results)) {
      return {
        success: true,
        results: response.results as T[],
      };
    }

    return {
      success: false,
      aborted: Boolean(response?.aborted ?? true),
      message: typeof response?.message === 'string' ? (response.message as string) : undefined,
    };
  }

  /**
   * Create a scoped view of the SDK where every command automatically carries the provided client id.
   */
  scope(clientId: string): TransactionScope {
    if (!clientId) {
      throw new Error('Transaction scope requires a clientId');
    }

    const baseClient = this.client;
    const scopedClient = {
      sendCommand: (command: string, payload: Record<string, unknown> = {}) => {
        const scopedPayload = { ...payload };
        if (!('client_id' in scopedPayload)) {
          scopedPayload.client_id = clientId;
        }
        return baseClient.sendCommand(command, scopedPayload);
      },
      ping: baseClient.ping.bind(baseClient),
      health: baseClient.health.bind(baseClient),
      close: baseClient.close.bind(baseClient),
    } as unknown as SynapClient;

    return {
      clientId,
      kv: new KVStore(scopedClient),
      hash: new HashManager(scopedClient),
      list: new ListManager(scopedClient),
      set: new SetManager(scopedClient),
      sortedSet: new SortedSetManager(scopedClient),
      queue: new QueueManager(scopedClient),
      stream: new StreamManager(scopedClient),
      pubsub: new PubSubManager(scopedClient),
      script: new ScriptManager(scopedClient),
      hyperloglog: new HyperLogLogManager(scopedClient),
    };
  }
}

File: sdks/rust/src/transactions.rs
//! Redis-compatible transaction support (MULTI/EXEC/WATCH/DISCARD)

use crate::client::SynapClient;
use crate::error::Result;
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

/// Options for transaction commands
#[derive(Debug, Clone, Default)]
pub struct TransactionOptions {
    pub client_id: Option<String>,
}

impl TransactionOptions {
    fn into_payload(self) -> Value {
        match self.client_id {
            Some(client_id) => json!({"client_id": client_id}),
            None => json!({}),
        }
    }
}

/// Standard response for MULTI/DISCARD/WATCH/UNWATCH
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransactionResponse {
    pub success: bool,
    #[serde(default)]
    pub message: Option<String>,
}

/// Result returned by EXEC
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum TransactionExecResult {
    Success {
        results: Vec<Value>,
    },
    Aborted {
        aborted: bool,
        #[serde(default)]
        message: Option<String>,
    },
}

/// Helper for sending raw commands within a transaction
pub struct TransactionCommandClient {
    client: SynapClient,
    client_id: String,
}

impl TransactionCommandClient {
    /// Send a raw command ensuring the `client_id` is attached
    pub async fn send_command(&self, command: &str, mut payload: Value) -> Result<Value> {
        if let Value::Object(ref mut map) = payload {
            map.insert(
                "client_id".to_string(),
                Value::String(self.client_id.clone()),
            );
        }

        self.client.send_command(command, payload).await
    }

    /// Access underlying client id
    pub fn client_id(&self) -> &str {
        &self.client_id
    }
}

/// Transaction manager exposing MULTI/EXEC/WATCH/DISCARD
#[derive(Clone)]
pub struct TransactionManager {
    client: SynapClient,
}

impl TransactionManager {
    pub(crate) fn new(client: SynapClient) -> Self {
        Self { client }
    }

    /// Start a transaction (MULTI)
    pub async fn multi(&self, options: TransactionOptions) -> Result<TransactionResponse> {
        let response = self
            .client
            .send_command("transaction.multi", options.clone().into_payload())
            .await?;
        Self::parse_response(response)
    }

    /// Discard an active transaction (DISCARD)
    pub async fn discard(&self, options: TransactionOptions) -> Result<TransactionResponse> {
        let response = self
            .client
            .send_command("transaction.discard", options.clone().into_payload())
            .await?;
        Self::parse_response(response)
    }

    /// Watch keys for optimistic locking (WATCH)
    pub async fn watch(
        &self,
        keys: &[impl AsRef<str>],
        options: TransactionOptions,
    ) -> Result<TransactionResponse> {
        let mut payload = options.into_payload();
        if let Value::Object(ref mut map) = payload {
            map.insert(
                "keys".into(),
                Value::Array(
                    keys.iter()
                        .map(|k| Value::String(k.as_ref().to_string()))
                        .collect(),
                ),
            );
        }

        let response = self
            .client
            .send_command("transaction.watch", payload)
            .await?;
        Self::parse_response(response)
    }

    /// Remove all watched keys (UNWATCH)
    pub async fn unwatch(&self, options: TransactionOptions) -> Result<TransactionResponse> {
        let response = self
            .client
            .send_command("transaction.unwatch", options.clone().into_payload())
            .await?;
        Self::parse_response(response)
    }

    /// Execute queued commands (EXEC)
    pub async fn exec(&self, options: TransactionOptions) -> Result<TransactionExecResult> {
        let response = self
            .client
            .send_command("transaction.exec", options.clone().into_payload())
            .await?;

        if response["results"].is_array() {
            let result = serde_json::from_value::<Vec<Value>>(response["results"].clone())?;
            return Ok(TransactionExecResult::Success { results: result });
        }

        let aborted = response["aborted"].as_bool().unwrap_or(true);
        let message = response["message"].as_str().map(|s| s.to_string());
        Ok(TransactionExecResult::Aborted { aborted, message })
    }

    /// Create a helper client that automatically injects `client_id` for raw commands
    pub fn command_client(&self, client_id: impl Into<String>) -> TransactionCommandClient {
        TransactionCommandClient {
            client: self.client.clone(),
            client_id: client_id.into(),
        }
    }

    fn parse_response(response: Value) -> Result<TransactionResponse> {
        let success = response["success"].as_bool().unwrap_or(true);
        let message = response["message"].as_str().map(|s| s.to_string());
        Ok(TransactionResponse { success, message })
    }
}
